src/System.Private.Xml/tests/XmlResolver/System.Xml.XmlResolver.Tests/XmlPreloadedResolverGetEntity.cs

```diff
@@ -81,17 +107,71 @@ public void XmlResolverGetEntityAsyncWithInvalidData()
+        {
+            var xmlResolver = new XmlPreloadedResolver(XmlKnownDtds.All);
+            foreach (var dtdInfo in GetKnownDtds())
+            {
+                string expectedContent = NormalizeContent(File.ReadAllText(dtdInfo.SourcePath));
+
+                var uri = new Uri(dtdInfo.PublicId, UriKind.RelativeOrAbsolute);
+                Stream streamResult = await xmlResolver.GetEntityAsync(uri, null, null) as Stream;
+                Assert.NotNull(streamResult);
+                byte[] data = new byte[streamResult.Length];
+                streamResult.Read(data, 0, Convert.ToInt32(streamResult.Length));
+                Assert.Equal(expectedContent, NormalizeContent(Encoding.ASCII.GetString(data)));
+
+                uri = new Uri(dtdInfo.SystemId, UriKind.RelativeOrAbsolute);
+                streamResult = await xmlResolver.GetEntityAsync(uri, null, null) as Stream;
+                Assert.NotNull(streamResult);
+                data = new byte[streamResult.Length];
+                streamResult.Read(data, 0, Convert.ToInt32(streamResult.Length));
+                Assert.Equal(expectedContent, NormalizeContent(Encoding.ASCII.GetString(data)));
+            }
+        }
+
+        private static IEnumerable<(string PublicId, string SystemId, string SourcePath)> GetKnownDtds()
+        {
+            var xhtmlDtds = new (string PublicId, string RelativeId)[]
+            {
+                ("-//W3C//DTD XHTML 1.0 Strict//EN", "xhtml1-strict.dtd"),
+                ("-//W3C//DTD XHTML 1.0 Transitional//EN", "xhtml1-transitional.dtd"),
+                ("-//W3C//DTD XHTML 1.0 Frameset//EN", "xhtml1-frameset.dtd"),
+                ("-//W3C//ENTITIES Latin 1 for XHTML//EN", "xhtml-lat1.ent"),
+                ("-//W3C//ENTITIES Symbols for XHTML//EN", "xhtml-symbol.ent"),
+                ("-//W3C//ENTITIES Special for XHTML//EN", "xhtml-special.ent"),
+            };
+            var rssDtds = new (string PublicId, string RelativeId)[]
+            {
+                ("-//Netscape Communications//DTD RSS 0.91//EN", "rss-0.91.dtd"),
+            };
+
+            string dtdFolderRoot = @"Utils\DTDs\";
+
+            return Enumerable.Concat(
+                GetKnownDtds(xhtmlDtds, "http://www.w3.org/TR/xhtml1/DTD/", Path.Combine(dtdFolderRoot, @"XHTML10\no_comments\")),
+                GetKnownDtds(rssDtds, "http://my.netscape.com/publish/formats/", Path.Combine(dtdFolderRoot, @"RSS091\no_comments\")));
+
+            IEnumerable<(string PublicId, string SystemId, string SourcePath)> GetKnownDtds(
+                IEnumerable<(string PublicId, string RelativeId)> ids,
+                string systemUrlPrefix,
+                string pathPrefix) =>
+                ids.Select(x => (x.PublicId, systemUrlPrefix + x.RelativeId, pathPrefix + x.RelativeId));
+        }
+
+        private static string NormalizeContent(string content) => content.Replace("\0", "").Replace("\r\n", "\n");
```

Comment:
why is `.Replace("\0", "")` needed?

Reply:
I don't know. This was in the code before. I know it looks like I added it, but it only appears that way because I moved it - this was a refactoring to avoid duplication.

Reply:
...although I've just realised that I didn't then refactor the original place I copied it from to use this - I only did that with the new places I added. Line 59 performs the same `Replace` without using this helper. That's the original one. Do you want me to modify it so that this also uses the same helper?

Reply:
yes please

Reply:
Done